src/SQLStore/ByIdDataRebuildDispatcher.php

```diff
@@ -0,0 +1,292 @@
+
+	/**
+	 * @var boolean
+	 */
+	private $useUpdateJobParseMode = true;
+
+	/**
+	 * @var boolean
+	 */
+	private $useJobQueue = true;
+
+	/**
+	 * @var array|false
+	 */
+	private $namespaces = false;
+
+	/**
+	 * @var integer
+	 */
+	private $iterationLimit = 1;
+
+	/**
+	 * @since 2.3
+	 *
+	 * @param Store $store
+	 */
+	public function __construct( Store $store ) {
+		$this->store = $store;
+	}
+
+	/**
+	 * @see MessageReporter::reportMessage
+	 *
+	 * @since 2.3
+	 *
+	 * @param string $message
+	 */
+	public function reportMessage( $message ) {
+		return '';
+	}
+
+	/**
+	 * @since 2.3
+	 *
+	 * @param boolean $updateJobParseMode
+	 */
+	public function setUpdateJobToUseParseMode( $updateJobParseMode ) {
+		$this->useUpdateJobParseMode = (bool)$updateJobParseMode;
+	}
+
+	/**
+	 * @since 2.3
+	 *
+	 * @param boolean $useJobQueueScheduler
+	 */
+	public function setUpdateJobToUseJobQueueScheduler( $useJobQueueScheduler ) {
+		$this->useJobQueueScheduler = (bool)$useJobQueueScheduler;
+	}
+
+	/**
+	 * @since 2.3
+	 *
+	 * @param array|false $namespaces
+	 */
+	public function setNamespacesTo( $namespaces ) {
+		$this->namespaces = $namespaces;
+	}
+
+	/**
+	 * @since 2.3
+	 *
+	 * @param integer $iterationLimit
+	 */
+	public function setIterationLimit( $iterationLimit ) {
+		$this->iterationLimit = (int)$iterationLimit;
+	}
+
+	/**
+	 * @since 2.3
+	 *
+	 * @return integer
+	 */
+	public function getMaxId() {
+
+		$db = $this->store->getConnection( 'mw.db' );
+
+		$maxByPageId = $db->selectField(
+			'page',
+			'MAX(page_id)',
+			'',
+			__METHOD__
+		);
+
+		$maxBySmwId = $db->selectField(
+			\SMWSql3SmwIds::tableName,
+			'MAX(smw_id)',
+			'',
+			__METHOD__
+		);
+
+		return max( $maxByPageId, $maxBySmwId );
+	}
+
+	/**
+	 * Dispatching of a single or a chunk of ids in either online or batch mode
+	 * using the JobQueueScheduler
+	 *
+	 * @since 2.3
+	 *
+	 * @param integer &$id
+	 */
+	public function dispatchRebuildFor( &$id ) {
+
+		$updatejobs = array();
+
+		// was nothing done in this run?
+		$emptyrange = true;
+
+		$this->createUpdateJobsForTitleIdRange( $id, $updatejobs );
+
+		if ( $updatejobs !== array() ) {
+			$emptyrange = false;
+		}
+
+		$this->createUpdateJobsForSmwIdRange( $id, $updatejobs, $emptyrange );
+
+		// Deprecated since 2.3, use 'SMW::SQLStore::BeforeDataRebuildJobInsert'
+		wfRunHooks('smwRefreshDataJobs', array( &$updatejobs ) );
+
+		Hooks::run( 'SMW::SQLStore::BeforeDataRebuildJobInsert', array( $this->store, &$updatejobs ) );
+
+		if ( $this->useJobQueueScheduler ) {
+			JobBase::batchInsert( $updatejobs );
+		} else {
+			foreach ( $updatejobs as $job ) {
+				$job->run();
+			}
+		}
+
+		// -1 means that no next position is available
+		$this->findNextIdPosition( $id, $emptyrange );
+
+		// Decimal between 0 and 1 to indicate the overall progress of the refreshing
+		return $id > 0 ? $id / $this->getMaxId() : 1;
+	}
+
+	private function createUpdateJobsForTitleIdRange( $id, &$updatejobs ) {
```

Comment:
Type hints missing. Then again, they where not there before, and talking small refactoring steps is good